conda_mirror/diff_tar.py

```diff
@@ -0,0 +1,191 @@
+    """
+    Asssuming the global `mirror_dir` is set, iterate all sub-directories
+    which contain a repodata.json and repodata.json.bz2 file.
+    """
+    for root, unused_dirs, files in os.walk(mirror_dir):
+        if 'repodata.json' in files and 'repodata.json.bz2' in files:
+            yield root
+
+
+def all_repodata():
+    """
+    Return a dictionary mapping all repository sub-directories to the conda
+    package list as respresented by the 'packages' field in repodata.json.
+    """
+    d = {}
+    for repo_path in find_repos():
+        with open(join(repo_path, 'repodata.json')) as fi:
+            index = json.load(fi)['packages']
+        d[repo_path] = index
+    return d
+
+
+def verify_all_repos():
+    """
+    Verify all the MD5 sum of all conda packages listed in all repodata.json
+    files in the repository.
+    """
+    d = all_repodata()
+    for repo_path, index in d.items():
+        for fn, info in index.items():
+            path = join(repo_path, fn)
+            if info['md5'] == md5_file(path):
+                continue
+            print('MD5 mismatch: %s' % path)
+
+
+def write_reference():
+    """
+    Write the "reference file", which is a collection of the content of all
+    repodata.json files.
+    """
+    data = json.dumps(all_repodata(), indent=2, sort_keys=True)
+    # make sure we have newline at the end
+    if not data.endswith('\n'):
+        data += '\n'
+    with open(reference_path, 'w') as fo:
+        fo.write(data)
+
+
+def read_reference():
+    """
+    Read the "reference file" from disk and return its content as a dictionary.
+    """
+    try:
+        with open(reference_path) as fi:
+            return json.load(fi)
+    except FileNotFoundError:
+        sys.exit('No such file: %s' % reference_path)
+
+
+def get_updates():
+    """
+    Compare the "reference file" to the actual the repository (all the
+    repodata.json files) and iterate the new and updates files in the
+    repository.  That is, the files which need to go into the differential
+    tarball.
+    """
+    d1 = read_reference()
+    d2 = all_repodata()
+    for repo_path, index2 in d2.items():
+        index1 = d1.get(repo_path, {})
+        if index1 != index2:
+            for fn in 'repodata.json', 'repodata.json.bz2':
+                yield relpath(join(repo_path, fn), mirror_dir)
+        for fn, info2 in index2.items():
+            info1 = index1.get(fn, {})
+            if info1.get('md5') != info2['md5']:
+                yield relpath(join(repo_path, fn), mirror_dir)
+
+
+def tar_repo(outfile='update.tar', verbose=False):
+    """
+    Write the so-called differential tarball, see get_updates().
+    """
+    t = tarfile.open(outfile, 'w')
+    for f in get_updates():
+        if verbose:
+            print('adding: %s' % f)
+        t.add(join(mirror_dir, f), f)
+    t.close()
+    if verbose:
+        print("Wrote: %s" % outfile)
+
+
+def main():
+    from optparse import OptionParser
```

Comment:
Woah, I thought this was deprecated in favor of argparse

Reply:
Oh, I guess it is deprecated now.  It's just what I've always been using.

Reply:
Haha That's OK!

Reply:
Switched to argparse 843b403a5f46ee815dccc7fd0ad5451867b2d643